---
title: "20201120_openASO_position"
author: "Kim Wellman, Lela Lackey"
date: "11/20/2020"
output:
  html_document: 
    code_folding: hide
---

### Purpose
This script takes a pre-processed data frame of gene ID, gene sequence, and the reverse compliment of the ASO sequence and maps start/stop positions of the ASO using transcriptomic coordinates that are then converted to genomic coordinates. The output of this script should be in BED file format.

### Setup

```{r setup, warning = FALSE}

# clear the global environment. 
rm(list=ls())

# get packages. 
suppressMessages(library(tidyverse))
library(readr)
library(Biostrings)
library(splitstackshape)
library(ensembldb)
library(EnsDb.Hsapiens.v86)
library(insect)
library(dplyr)

```

### Data input: pre-processed data from Ivan and Axel

```{r import, echo = TRUE, warning = FALSE}

genseq_byaso <- read.csv("~/ASO_position_script/Complete_ASOtoTranscriptSeq.tsv", sep="", stringsAsFactors=FALSE)

# generate the reverse complement of the ASOs.
genseq_byaso$aso_rc_seq <- rc(genseq_byaso$ASOseq)

# confirm geneID of isoform to transcript_id
genseq_byaso$transcript_id <- gsub("\\..*","",genseq_byaso$hg38.knownGene.name)

```

### Map the transcriptomic coordinates

These coordinates are relative to the first nucleotide of the transcript (not the CDS) in order to be compatible with the 'transcriptToGenome' ensembldb functionality.

```{r transcriptomic coordinates, warning = FALSE}

# loop through each row to look for where each ASO sequence is located along the target transcript sequence. The'start' and 'end' positions are in relation to the start of the transcript and also represent character position within each transcript sequence string. 

aso_txt_pos <- genseq_byaso
aso_txt_pos$start <- NA
aso_txt_pos$end <- NA

for(i in 1:dim(aso_txt_pos)[1]){
  string <- genseq_byaso[i, "Sequence"]
  pattern <- genseq_byaso[i, "aso_rc_seq"]
  coord <- as.data.frame(str_locate(string, pattern))
  aso_txt_pos[i, "start"] <- coord[1, "start"]
  aso_txt_pos[i, "end"] <- coord[1, "end"]
}

aso_txt_pos <- na.omit(aso_txt_pos, cols = "start")

```

### Convert the transcriptomic coordinates into genomic coordinates


```{r genomic coordinates, warning = FALSE}

# convert the transcript position information into an IRanges format. 
# ir <- IRanges(start = aso_txt_pos$start, end = aso_txt_pos$end, width = (aso_txt_pos$end - aso_txt_pos$start + 1), names = aso_txt_pos$transcript_id)

ir <- IRanges(start = aso_txt_pos$start, width = (aso_txt_pos$end - aso_txt_pos$start + 1), names = aso_txt_pos$transcript_id)

# create an EnsDB object containing genomic position information.
dbfile <- system.file("extdata/EnsDb.Hsapiens.v86.sqlite", package = "EnsDb.Hsapiens.v86")
db <- EnsDb(dbfile)

gr <- transcriptToGenome(ir, db)

```

```{r GRanges to BED format, warning = FALSE}

# create dataframe from GrangesList which groups GRanges together, ASOs by transcript. Loci.group connects multiple ranges (across exons) back to the original ASO. 
# in order to connect the groups (potentially multiple GRanges when ASOs cross exons), we need the corresponding txt start/stop. 

gnm_df <- data.frame(loci = seqnames(gr), 
  starts = start(gr) - 1,
  ends = end(gr), gr@unlistData@elementMetadata@listData[["tx_start"]], gr@unlistData@elementMetadata@listData[["tx_end"]]) 

colnames(gnm_df)[colnames(gnm_df) == "gr.unlistData.elementMetadata.listData...tx_start..."] <- "tx_start"
colnames(gnm_df)[colnames(gnm_df) == "gr.unlistData.elementMetadata.listData...tx_end..."] <- "tx_end"

gnm_df <- gnm_df %>% dplyr::select(loci.group_name, loci.value, loci.group, starts.value, ends.value, tx_start, tx_end)

final_df <- full_join(aso_txt_pos, gnm_df, by = c('transcript_id' = 'loci.group_name', 'start' = 'tx_start', 'end' = 'tx_end'))

# reshape the data by the loci.group

# convert the data frame to bed format.

# write.csv(final_df, file = "openASO_gr.bed", quote = F, sep = "\t", row.names = F, col.names = F)
```




```{r}
sessionInfo()

```
